app/src/main/java/com/worldsnas/starplayer/ExoPlayerService.kt

```diff
@@ -0,0 +1,684 @@
+        }
+    }
+
+    override fun onDestroy() {
+        super.onDestroy()
+        releasePlayer()
+        if (mAudioManager != null) mAudioManager!!.abandonAudioFocus(this)
+        sendBroadcast_stateIdle()
+    }
+
+    override fun onAudioFocusChange(focusChange: Int) {
+        if (focusChange >= 0 && (pausedByAudioListener || player!!.playWhenReady)) startPlayer() else {
+            if (player!!.playWhenReady) pausedByAudioListener = true
+            pausePlayer()
+        }
+    }
+
+    private fun resetDataQueue() {
+        if (trackModelsQueue == null) trackModelsQueue =
+            ArrayList() else trackModelsQueue!!.clear()
+    }
+
+    private fun play(trackModel: Music?) {
+        val trackModels = ArrayList<Music?>()
+        trackModels.add(trackModel)
+        play(trackModels)
+    }
+
+    private fun play(trackModels: ArrayList<Music?>?) {
+        reInit()
+        if (trackModels != null) {
+            trackModelsQueue = trackModels
+            val mediaSources =
+                arrayOfNulls<MediaSource>(trackModels.size)
+            for (i in trackModels.indices) {
+                mediaSources[i] =
+                    buildMediaSource(Uri.parse(trackModels[i]!!.address))
+            }
+            mediaSource = ConcatenatingMediaSource(*mediaSources)
+
+            player!!.prepare(mediaSource!!)
+            startPlayer()
+        }
+    }
+
+    private fun add(trackModel: Music?) {
+        if (trackModelsQueue == null || trackModelsQueue!!.size == 0) play(
+            trackModel
+        ) else {
+            addTrackIfNotAdded(trackModel)
+        }
+    }
+
+    private fun addTrackIfNotAdded(trackModel: Music?) {
+        if (!isInCurrentQueue(trackModel)) {
+            mediaSource!!.addMediaSource(buildMediaSource(Uri.parse(trackModel!!.address)))
+            trackModelsQueue!!.add(trackModel)
+        }
+    }
+
+    private fun add(trackModels: ArrayList<Music?>) {
+        if (trackModelsQueue == null || trackModelsQueue!!.size == 0) play(
+            trackModels
+        ) else {
+            for (i in trackModels.indices) addTrackIfNotAdded(trackModels[i])
+        }
+    }
+
+    private fun isInCurrentQueue(trackModel: Music?): Boolean {
+        for (model in trackModelsQueue!!) if (model!!.id == trackModel!!.id) return true
+        return false
+    }
+
+    private fun pausePlayer() {
+        playbackPosition = player!!.currentPosition
+        currentWindow = player!!.currentWindowIndex
+        player!!.playWhenReady = false
+        player!!.playbackState
+        showCurrentTrackNotification()
+    }
+
+    private fun startPlayer() {
+        pausedByAudioListener = false
+        player!!.seekTo(currentWindow, playbackPosition)
+        player!!.playWhenReady = true
+        player!!.playbackState
+        showCurrentTrackNotification()
+    }
+
+    private fun initializePlayer() {
+        if (player == null) {
+            player = ExoPlayerFactory.newSimpleInstance(
```

Comment:
in kaht hm barresi konid lotfan deprecate shode

Reply:
ok shod